File: haskell/parser.hs
-- Parser for math expressions based on Abstract Syntax Tree (AST)
import Data.Bits (xor)

data Expr = Add Expr Expr   -- Sum
    | Sub Expr Expr -- Subtraction
    | Mul Expr Expr -- Multiplication
    | Div Expr Expr -- Division
    | Pow Expr Expr -- Power
    | And Expr Expr -- Logic and
    | Or Expr Expr  -- Logic or
    | Xor Expr Expr  -- Logic xor
    | Not Expr  -- Login not
    | Val Double    -- Numeric value
    | BoolVal Bool  -- Boolean value
    deriving (Show)

-- Hierarchical parser: values, terms (*,/,^) and expressions (+,-)
-- Extracts values or parenthesis
parseValue :: String -> (Expr, String)
parseValue ('-':xs) =
    let (Val num, rest) = parseValue xs -- Support for negative numbers
    in (Val (-num), rest)
parseValue ('(':xs) =
    let (expr, rest) = parseLogic xs
    in case rest of
        (')':rest') -> (expr, rest')    -- Parenthesis close
        _ -> error "Error: missing ')' "
parseValue ('T':'r':'u':'e':xs) = (BoolVal True, xs)    -- Get True value
parseValue ('F':'a':'l':'s':'e':xs) = (BoolVal False, xs)   -- Get False value
parseValue xs =
    let (num, rest) = span (`elem` "0123456789.") xs    -- Split xs in (xs1, xs2) where xs1 contains the first elemnts of xs that are digits or '.'
    in if null num 
        then error "Error: numeric value expected"
        else (Val (read num), rest) -- 'read' convert the num from string to Num value

-- Parse power (^)
parsePow :: String -> (Expr, String)
parsePow input =
    let (base, rest) = parseValue input
    in parsePow' base rest
    where
        parsePow' left ('^':xs) =
            let (right, rest) = parseValue xs
            in parsePow' (Pow left right) rest
        parsePow' left right = (left, right)    -- No other operator

-- Terms parsing (*, /) 
parseTerm :: String -> (Expr, String)
parseTerm input =
    let (factor, rest) = parsePow input
    in parseTerm' factor rest 
    where
        parseTerm' left ('*':xs) =
            let (right, rest) = parsePow xs
            in parseTerm' (Mul left right) rest
        parseTerm' left ('/':xs) =
            let (right, rest) = parsePow xs
            in parseTerm' (Div left right) rest
        parseTerm' left rest = (left, rest) -- No other operator

-- Expressions parsing
parseExpr :: String -> (Expr, String)
parseExpr input =
    let (term, rest) = parseTerm input
    in parseExpr' term rest
    where
        parseExpr' left ('+':xs) =
            let (right, rest) = parseTerm xs
            in parseExpr' (Add left right) rest
        parseExpr' left ('-':xs) =
            let (right, rest) = parseTerm xs
            in parseExpr' (Sub left right) rest
        parseExpr' left rest = (left, rest) -- No other operator

-- Logical expressions parsing
parseLogic :: String -> (Expr, String)
parseLogic ('N':'o':'t':xs) =
    let (expr, rest) = parseLogic xs
    in (Not expr, rest)
parseLogic input = 
    let (expr, rest) = parseExpr input
    in parseLogic' expr rest
    where
        parseLogic' left ('&':'&':xs) =
            let (right, rest) = parseExpr xs
            in parseLogic' (And left right) rest
        parseLogic' left ('|':'|':xs) =
            let (right, rest) = parseExpr xs
            in parseLogic' (Or left right) rest
        parseLogic' left ('X':'o':'r':xs) =
            let (right, rest) = parseExpr xs
            in parseLogic' (Xor left right) rest
        parseLogic' left right = (left, right)

-- Parsing function
parse :: String -> Expr
parse input =
    let (expr, rest) = parseLogic (filter (/=' ') input) -- Remove blank spaces
    in if null rest 
        then expr
        else error ("Error: missing values " ++ rest)

-- Evaluate AST for math expr
eval :: Expr -> Double
eval (Val x) = x    -- A single value
eval (Add e1 e2) = eval e1 + eval e2    -- Sum of sub-tree results
eval (Sub e1 e2) = eval e1 - eval e2    -- Subtraction of sub-tree results
eval (Mul e1 e2) = eval e1 * eval e2    -- Multiplication of sub-tree results
eval (Div e1 e2) = 
    let denom = eval e2
    in if denom == 0
        then error "Error: zero division!"
        else eval e1 / denom    -- Division of sub-tree results
eval (Pow e1 e2) = eval e1 ** eval e2    -- Power

-- Evaluate AST for bool expr
evalBool :: Expr -> Bool
evalBool (BoolVal b) = b    -- A single value
evalBool (And e1 e2) = evalBool e1 && evalBool e2
evalBool (Or e1 e2) = evalBool e1 || evalBool e2
evalBool (Not e1) = not(evalBool e1)
evalBool (Xor e1 e2) = evalBool e1 `xor` evalBool e2
evalBool _ = error "Error: boolean operation or non-boolean value"

main::IO ()
main = do
    putStrLn "Insert a math expression, ex. 3 + 5 * (2 - 8) or bool expression, ex. True && True"
    input <- getLine
    let ast = parse input
    putStrLn "AST: "
    print ast
    putStrLn "Result:"
    case ast of
        BoolVal _ -> print(evalBool ast)
        And _ _ -> print(evalBool ast)
        Or _  _-> print(evalBool ast)
        Xor _ _ -> print(evalBool ast)
        Not _ -> print(evalBool ast)
        _ -> print(eval ast) 
      

File: rust/parser.rs
use std::iter::Peekable;

#[derive(Debug, Clone)]
enum Expr {
    Add(Box<Expr>, Box<Expr>), // Sum
    Sub(Box<Expr>, Box<Expr>), // Subtraction
    Mul(Box<Expr>, Box<Expr>), // Multiplication
    Div(Box<Expr>, Box<Expr>), // Division
    Pow(Box<Expr>, Box<Expr>), // Power
    Not(Box<Expr>), // Logical not
    And(Box<Expr>, Box<Expr>),  // Logical and
    Or(Box<Expr>, Box<Expr>),   // Logical or
    Xor(Box<Expr>, Box<Expr>),  // Logical xor
    Val(f64),                  // Numeric value
}

impl Expr {
    fn eval(&self) -> f64 {
        match self {
            Expr::Val(x) => *x,
            Expr::Add(e1, e2) => e1.eval() + e2.eval(),
            Expr::Sub(e1, e2) => e1.eval() - e2.eval(),
            Expr::Mul(e1, e2) => e1.eval() * e2.eval(),
            Expr::Div(e1, e2) => {
                let denom = e2.eval();
                if denom == 0.0 {
                    panic!("Error: zero division!");
                }
                e1.eval() / denom
            }
            Expr::Pow(e1, e2) => e1.eval().powf(e2.eval()),
            Expr::Not(e) => {
                if e.eval() != 0.0 {
                    0.0
                } else {
                    1.0
                }
            },
            Expr::And(e1, e2) => {
                if e1.eval() != 0.0 && e2.eval() != 0.0 {
                    1.0
                } else {
                    0.0
                }
            },
            Expr::Or(e1, e2) => {
                if e1.eval() != 0.0 || e2.eval() != 0.0 {
                    1.0
                } else {
                    0.0
                }
            },
            Expr::Xor(e1, e2) => {
                if (e1.eval() != 0.0) ^ (e2.eval() != 0.0) {
                    1.0
                } else {
                    0.0
                }
            },
        }
    }
}

// Had to specify the mut ref type as an Iterator. The function accepts any valid iterator
fn parse_value<I>(chars: &mut Peekable<I>) -> Expr
where
    I: Iterator<Item = char>
{
    if let Some(&'-') = chars.peek() {
        chars.next();
        match parse_value(chars) {
            Expr::Val(num) => Expr::Val(-num),
            _ => panic!("Error: numeric value expected after '-'"),
        }
    } else if let Some(&'!') = chars.peek() {   // Logical not
        chars.next();
        let expr = parse_value(chars);
        Expr::Not(Box::new(expr))

    } else if let Some(&'(') = chars.peek() {
        chars.next();
        let expr = parse_expr(chars);
        if chars.next() != Some(')') {
            panic!("Error: missing ')' ");
        }
        expr
    } else {
        let mut num_str = String::new();
        while let Some(&ch) = chars.peek() {
            if ch.is_numeric() || ch == '.' {
                num_str.push(ch);
                chars.next();
            } else {
                break;
            }
        }
        if num_str.is_empty() {
            panic!("Error: numeric value expected");
        }
        Expr::Val(num_str.parse::<f64>().unwrap())
    }
}

fn parse_pow<I>(chars: &mut Peekable<I>) -> Expr
where
    I: Iterator<Item = char>
{
    let base = parse_value(chars);
    let mut result = base;
    while let Some(&'^') = chars.peek() {
        chars.next();
        let exponent = parse_value(chars);
        result = Expr::Pow(Box::new(result), Box::new(exponent));
    }
    result
}

fn parse_term<I>(chars: &mut Peekable<I>) -> Expr
where
    I: Iterator<Item = char>
{
    let factor = parse_pow(chars);
    let mut result = factor;
    while let Some(&op) = chars.peek() {
        if op == '*' || op == '/' {
            chars.next();
            let next_factor = parse_pow(chars);
            result = match op {
                '*' => Expr::Mul(Box::new(result), Box::new(next_factor)),
                '/' => Expr::Div(Box::new(result), Box::new(next_factor)),
                _ => panic!("Unexpected operator: {}. Only '*' or '/' are allowed at this stage", op),
            };
        } else {
            break;
        }
    }
    result
}

fn parse_logic<I>(chars: &mut Peekable<I>) -> Expr
where
    I: Iterator<Item = char>
{
    let expr = parse_term(chars);
    let mut result = expr;
    while let Some(&op) = chars.peek() {
        if op == '&' || op == '|' || op == 'x' {    // x for Xor
            chars.next();
            let next_expr = parse_term(chars);
            result = match op {
                '&' => Expr::And(Box::new(result), Box::new(next_expr)),
                '|' => Expr::Or(Box::new(result), Box::new(next_expr)),
                'x' => Expr::Xor(Box::new(result), Box::new(next_expr)),
                _ => panic!("Unexpected operator: {}. Only '&' '|' '^' are allowed at this stage", op),
            };
        } else {
            break;
        }
    }
    result
}  

fn parse_expr<I>(chars: &mut Peekable<I>) -> Expr
where
    I: Iterator<Item = char>
{
    let term = parse_logic(chars);
    let mut result = term;
    while let Some(&op) = chars.peek() {
        if op == '+' || op == '-' {
            chars.next();
            let next_term = parse_term(chars);
            result = match op {
                '+' => Expr::Add(Box::new(result), Box::new(next_term)),
                '-' => Expr::Sub(Box::new(result), Box::new(next_term)),
                _ => panic!("Unexpected operator: {}. Only '+' or '-' are allowed at this stage", op),
            };
        } else {
            break;
        }
    }
    result
}

fn parse(input: &str) -> Expr {
    let mut chars = input.chars().filter(|&c| !c.is_whitespace()).peekable();
    let expr = parse_expr(&mut chars);
    if chars.peek().is_some() {
        panic!("Error: unexpected characters after parsing");
    }
    expr
}

fn main() {
    use std::io::{self, Write};

    println!("Insert a math expression, ex. 3 + 5 * (2 - 8) or 1 & !0");
    let mut input = String::new();
    io::stdout().flush().unwrap();
    io::stdin().read_line(&mut input).unwrap();

    let input = input.trim();
    let ast = parse(input);

    println!("AST: {:?}", ast);
    println!("Result: {}", ast.eval());
}
